refactor: Posting Journal and Associated Tools

jniles opened the pull request:
This Pull Request finalizes the refactor of the Posting Journal and associated modules.  The commit history contains a full list of changes.  In brief:
 1. Editing has been moved out into a modal.  By breaking out the editing code, the complexity of the Posting Journal is reduced.  It also enforces editing a single transaction at a time.
 2. The Trial Balance has moved into Stored Procedures.  This separates the JS code from the SQL code and speeds up the execution of the Trail Balance slightly.
 3. The Trial Balance no longer reloads data between `$state` changes.  The modal is much more snappy than previous.
 4. Links to the receipts/documents/patients associated with each transaction are embedded in the Posting Journal.  This makes it easy to find the details associated with a given transaction.
 5. The Posting Journal is able to load "posted" transactions.  To accommodate this, the Posting Journal footer has been improved to show the number of unposted or posted transactions.  Additionally, the "Posted Journal" module has been removed.

There have been a number of miscellaneous bug fixes addressed as well.  As with any change, this may introduce new bugs specific to the changes listed above.

Partially addresses #1432.

Closes #1034.
Closes #1163.
Closes #1402.
Closes #1500.
Closes #1640.
Closes #1659.
Closes #1716.
Closes #1717.
Closes #1724.
Closes #1832.
Closes #1839.
Closes #1921.
Closes #1928.
Closes #1934.
Closes #1943.
Closes #1944.
Closes #1948.
Closes #1950.
Closes #1961.
Closes #2031.
Closes #2041.

---

### TODO Before Merge
 - [x] Ensure all keys are translated into both French and English
 - [x] Ensure tests pass
 - [x] Make sure all filters/links work on Posting Journal page.
 - [x] Make sure all filters/links work on the Trial Balance
 - [x] Ensure all editors work as expected.
 - [x] Ensure that new code passes lint checks.

jniles committed:
fix(trial balance): selectively show errors

This commit fixes the show/hide errors buttons so that it is only
present when there are errors to correct.

jniles committed:
refactor(trial balance): use standard colors

This commit migrates the Trial Balance to using standardized colors for
the ui-grid on success/error states. These colors are based off the
bootstrap colors already ubiquitous throughout the application.

Closes #1717.

jniles committed:
refactor(trial balance): view details in Journal

This commit refactors the trial balance to show all the trial balance
details in the Journal, rather than trying to render a grouped ui-grid
on top of another grouped ui-grid.

Currently, this has some drawbacks:
 1. Because the filtering is done on the server, this requires an
 additional round-trip to the server. We could reasonably easily
 implement an inline-filter to take care of this, but it isn't a pattern
 we've followed so far.
 2. Destroys the previous filters; this may not be necessarily
 understood by the user.  We can revert this behavior with an
 angular.merge() of the two filters (cached and $state defined).

Partially addresses #1163.  Closes #1716.

sfount committed:
feat(journal) utility menu editing button

This commit adds a dedicated editing button that responds to the
current journals selection. The following states are supported:
No transactions selected - disabled
One transactions selected - enabled, allows editing
One Plus transactions selected - enabled, warning message

sfount committed:
feat(journal) add grid selection service

This commit adds a service that wraps common
use cases for Grid selection. Many different grids
have a need of knowing how many rows are selected
within logical groups (for example the number of
transactions selected given any number of rows).

jniles committed:
fix(journal): improve grid footer readability

This commit improves the Posting Journal's grid footer readability by
translating the full string and providing extra padding around the text.

sfount committed:
feat(journal) modal transaction editing infrastructure

This commit introduces the files required for opening any
transaction in a modal editor. Any record UUID can be passed to the
modal controller from any controller and the transaction information
is requested through the Jounral service

jniles committed:
refactor(db): Add stored proc. for Trial Balance

This commit adds stored procedures for the Trial Balance to the
procedures.sql file.  They all rely in preprocessing transactions to
store the record_uuids in a temporary table.  This should give a
performance improvement as well as increased clarity to the code
(separation of JS concerns versus SQL concerns).

sfount committed:
feat(journal) editing ui grid transaction rows

This commit introduces row editing for shared and individual row
values. UI Grid rows directly edit the underlying data and shared
values are applied to every grid row.

Objects are set up for tracking added, removed and edited rows.

sfount committed:
feat(journal) error handling message infrastructure

This commit adds the templates and basic logic for both
error handling infrastructure and presenting transactions
in a read only view.

sfount committed:
feat(journal) format edits and write save transactions

This commit updates the modal code to format edits as well as added and
removed rows and write these changes to the server. If errors are returned
they are displayed in the modal.

jniles committed:
refactor(Trial Balance): combine TB APIs into one

BREAKING CHANGE The Trial Balance APIs have now been combined into a
single route.

By combining routes, we simplify the number of HTTP requests sent for the
Trial Balance information, and allows the client to maintain stateful
information about the errors as well as the summary information.  It
further simplifies the parsing of information sent from the client -
everything is contained in a temporary table shared in the database
transaction.

jniles committed:
fix(journal): ensure debits/credits sort as number

This commit fixes a long-standing bug where the debits and credits in
the Posting Journal are treated as strings instead of numbers.  It also
updates the Trial Balance initialisation style to work with the new, in
development Trial Balance format.

jniles committed:
fix(csv): ensure exporter link is hidden

The exporter link is used only in JavaScript as a handle to grab the
underlying element.  In this case, the element should be invisible to
the client.  This commit sets the display to be none accordingly.

sfount committed:
feat(journal) update underlying data set on changes

This commit updates the journal to using a store - this allows
the removal and reinsertion of updated results given a successful edit.

This commit also adds loading indicators to the transaction editing
modal.

sfount committed:
refactor(journal) remove legacy in-line editing code

This commit refactors the journal and core grid services to rely
on the new feature modal editing and removes legacy structur required
to edit in-line.

It also updates the optimistic updates on the Journal module with a
visual warning for not being able to not having enough information to
update the client (with a prompt to refresh the data set as it may be
out of date).

sfount committed:
refactor(journal) read only infrastructure; translations

This commit ensures that the read only flag is respected by the different
components in the edit transaction modal. It also removes any hard
coded text elements in place of translatable keys.

sfount committed:
refactor(links) Implement reference link as a component

This commit removes the current implementation of a UI grid cell template
link and replaces it with a component. This component allows the programmer
to set a reference and a display value that can be shown to the user.

Advantages of a component design in this case:
* Can be used interchangably in and out of UI grids
* Can factor in global settings to link internally or redirect (documents
  vs. search
* Reads more naturally in markup
* Optionally display different text to the user than the reference
* Link behaviours can be updated in one central file rather than at every
  implementation

sfount committed:
refactor(links) Apply bh-reference-links to journal entities

This commit applies the `bh-reference-links` to all referenced
entities in the journal module.

jniles committed:
fix(i18n): align Posting Journal en.json

sfount committed:
feat(journal) include both posted and nonposted transactions

This commit updates the queries for the journal on the server using
a union between two known methods to combine substets. It also removes
aggregation logic that was used in legacy journal editing.

jniles committed:
chore(journal): merge and update latest journal

This commit update the Trial Balance's journal branch with @sfount's
changes, combining the journal.js file and removing duplicated effort
where possible.

jniles committed:
refactor(Trial Balance): rewrite Trial Balance

This commit completely rewrites the Trail Balance removing the old
multi-controller headers and replacing it with a single modal controller
and substate controllers for the modal body.  It also introduces the
Trial Balance Service that interacts with the server to load the Trial
Balance.

Key Differences/Features:
 1. Data is loaded a single time - no additional requests are made for
 data-by-account.
 2. Instantaneously switching between error and overview states.
 3. More room on the UI grid display
 4. ui-grid complexity reduce by removing grid grouping and features.
 5. [IN PROGRESS] Links to filter the journal based on the account and
 transaction.

Since all data is loaded on startup, it should remove all issues
reported in #1839.

Closes #1716. Closes #1163. Closes #1839.

jniles committed:
fix(Trial Balance): ensure grid export works

This commit ensures the export method is called when the export link is
clicked on the Trial Balance.

jniles committed:
feat(components): bhJournalLink for linkin records

This commit adds a component for linking record_uuids in the journal via
some display method.  Currently the display is the same one that is
shown as the display value of the filter - this may need to change.

A demonstration of this is in the Trial Balance Error Links.

jniles committed:
perf(Trial Balance): rewrite PostToGeneralLedger

This commit rewrites the PostToGeneralLedger function for speed.  It now
shares common functions with the Trial Balance to stage transactions.

Closes #1928.

jniles committed:
fix(tests): update Trial Balance integration tests

This commit updates the Trial Balance integration tests to the latest
API.

sfount committed:
feat(journal) add footer row transaction breakdown

This commit visuaully differentiates between posted and non-posted
transactions. It also adds a footer row that displays the following totals:
* number of rows
* number of posted transactions
* number of non-posted transactions
* number of remaining system non-posted transactions not shown

sfount committed:
feat(journal) show posted flag added to journal search

This commit adds an additional flag to the journal search modal
and updates the server side API to not show posted transactions by
default.

sfount committed:
fix(journal) ensure that the limit is respected on transaction search

This commit implements a work around for limiting the union of the
posting jouranl and general ledger.

sfount committed:
fix(links) reference links component updates on change

This commit ensures that the reference link component works in a UI grid
when underlying data is changed. UI Grid optimises the creation of components by
updating existing components instead of replacing them. This requires the
component to act and change the display label on change - it cannot assume
the reference value passed will be static.

sfount committed:
fix(journal) block journal editing of posted transactions

This commit adds a condition to posting journal editing to stop any
editing request on already posted requests. In the future super user
editing may need to be factored in here.

sfount committed:
refactor(journal) client side validation; editing posted records

This commit verifies on the client if the transaction to be edited is
marked as posted. This allows the application to stop the user before trying
to edit a posted record, the server will also respond with an invalid message if
this is circumvented.

sfount committed:
refactor(journal) Notify on transaction edit; success and warn

This commit adds success and warn state notifications to transaction editing.
If a transactions was edited a success notification will be displayed and the
logic is adjusted so that if no changes were made a warn notification will be
shown and the selected of the Journal UI Grid will remain unchanged.

sfount committed:
refactor(journal) transaction editing API delete all rows case

This commit handles the delete all rows case for the server transaction
editing API. If the client tools send removed rows that are greater
than the number of initial rows this will now result in a 400.

This should eventually encourage developers and then users to use the
proposed delete transaction API.

jniles committed:
test(journal): update tests for journal changes

This commit fixes the integration tests so that they pass with the
recent changes to the journal API.

jniles committed:
fix(journal): exchange debit/credit fields on edit

This commit updates the debit/credit fields to compute their values
based on the exchange rate of the day.  This makes sure the debit/credit
values are always up to date with their debit_equiv/credit_equiv
counterparts.

sfount committed:
test(journal) update transaction editing tests for modal editing

This commit updates the E2E that validate transaction editing to use the
latest modal editing structure. It also adds required IDs or data-methods to
transaction editing modal to allow tests to interact with components.

jniles committed:
fix(trialbalance): render currencies in footer

This commit makes sure that the footer of the Trial Balance does not
suffer from JS rounding flaws.

sfount committed:
fix(links) handle `documentPath` not defined case

This commit adds an additional check to the reference link server
controller; a redirect is not requested unless the document type is
defined and it specifies a documentPath - this stops an infinite
redirect in certain cases.

It also updates the patient document ('PA') to link to the patient
financial history document given a reference lookup.

sfount committed:
refactor(journal) use reference links on edit grid

This commit adds the `bh-reference-links` component for the `hrEntity`
field on the edit transaction grid.

It also enables sorting by number on debits and credits for the same
grid.

jniles committed:
fix(journal): right-align currency columns

This commit makes sure users can quickly compare currency values by
right-aligning their values.

jniles committed:
fix(journal): block Trial Balance for posted txns

This commit blocks the Trial Balance with a warning to the user if the
user has selected any posted records to perform a Trial Balance.  The
Trial Balance only makes sense on unposted records.

sfount committed:
test(links) update integration test for link route

This commit updates the patients latest invoice route in the integration
tests to use the patients uuid instead of the debtor uuid.

sfount committed:
refactor(journal) Inverse default transaction sort

This commit sorts transactions from smallest to largest by default.
Usually transactions with a smaller ID will have an earlier timestamp -
this change will implicitly (roughly) order transactions by date so that
the ordering is not swapped when moving to the grouped view.

lomamech committed:
refactor(Editing transaction Dates)
- This PR makes it possible to put the period as well as the year whenever
the date changes and prevent the modification of the date for closed or
inexistent fiscal years
- Implement of function lookupFiscalYearByDate for check the FiscalYear
  and period for one Date
-

lomamech committed:
refactor(Editing transaction Date)
- Fix the comment for the description of function
- Put all alteration of data in the function transformColumns()
- rename the name of function selectTransdate

jeremielodi committed:
feat(posting jour) marking transactions as comment

jniles committed:
test(journal): fix erroneous test data

This commit makes sure that the posting journal's data is in the correct
period.

jniles committed:
fix(i18n): translate journal modules

This commit rounds out the posting journal tools by translating all the
keys that are used.  It also prunes the `posting_journal.json` file to
find keys that are not used.

jniles committed:
chore(journal): journal.js passes lint checks

This reworks journal.js so that it passes lint checks.  It also removes
many of the old imports.

jniles committed:
fix(journal): fiscal year assigned to new lines

The previous journal edit code only considered edited lines to update
the fiscal_year_id on.  This behavior threw errors when new lines were
added to a transaction.  The errors are fixed by updating the editing
code to work on both new lines and old lines.

jniles committed:
chore(editModal): clean up lint issues

This commit cleans up the lint issues in the transaction edit modal of
the Posting Journal.

jniles committed:
fix(trial balance): pass account label to filters

This commit improves the UX by passing the full account label (human
readable label) to the Posting Journal filters.  The account is now
properly shown instead of just the number.

jniles committed:
fix(journal): right-align footer currencies

Fixes the currencies to be right-aligned instead of left-aligned.  It
also ensures footer values are defined before rendering "NaN" and
defaults to "0".

jniles committed:
fix(journal): broken comment reference

This commit fixes a broken reference that prevented commenting the
Posting Journal.

jniles committed:
fix(journal): right align edit modal footer

This commit right-aligns the currencies in the edit modal footer.

jniles committed:
fix(journal): allow PDF printing of posted txns

This commit fixes the PDF report by combining posted and unposted report
routes into one and the same route.

jniles committed:
fix(journal): misc errors

This commit fixes various miscellaneous usability errors and typos in
the journal, trial balance, and transaction editing process.
Specifically:
 1. Better French translation in the Journal Search Modal (thanks
 @mbayopanda)
 2. Description is automatically copied onto new modal line when adding
 a new lint (thanks @lomamech)
 3. Caught a typo that prevented commenting from working (@mbayopanda).
 4. Caught a bug that prevents editing a row when clicking on its group
 header (@lomamech).

jniles committed:
feat(journal): add transaction validation

This commit adds "offline" validation to the transaction modal, which
was the no 1 requested item on the feedback from developers.  The intent
is not to be comprehensive, but to prevent users from making silly
mistakes.

jniles committed:
test(journal): update editing tests

This commit updates the journal with editing tests for transactions.
These tests ensure that the validation works in the modal.

jniles committed:
fix(journal): comments can be removed

This commit allows the journal to remove comments.  Additionally, the
comment modal has been improved such that, if a unique value is used for
a comment, it is displayed in the input.  This makes the component feel
more like it belongs on the page.

jniles committed:
chore(journal); rename Posting Journal to Journal

This commit renames the Posting Journal and subsequent translations to
"Journal".  It also removes "Posted Journal" from the database.

Closes #2031

mbayopanda commented on client/src/js/services/grid/GridSelection.js:
```diff
@@ -0,0 +1,71 @@
+    // special treatment to header rows:
+    // if the header is selected, select all children.
+    // if the header is unselected, unselect all children.
+    if (isHeaderRow(row)) {
```
Good comments and it works well 👍

mbayopanda commented on client/src/modules/account_statement/modals/comment.modal.html:
```diff
@@ -12,9 +9,8 @@
-      ng-class="{ 'has-error' : ModalForm.$submitted && ModalForm.comment.$invalid }">
-      <input type="text" class="form-control" name="comment" ng-model="$ctrl.comment" required>
+    <div class="form-group" ng-class="{ 'has-error' : ModalForm.$submitted && ModalForm.comment.$invalid }">
+      <input type="text" class="form-control" name="comment" ng-model="$ctrl.comment">
```
Good fix 👍

mbayopanda commented on client/src/modules/general-ledger/general-ledger.ctrl.js:
```diff
@@ -1,239 +0,0 @@
-
-    Export.download(url, params, 'POSTING_JOURNAL.TITLE');
-  };
-}
```
:+1

mbayopanda commented on client/src/modules/general-ledger/general-ledger.html:
```diff
@@ -1,80 +0,0 @@
-    </div>
-  </div>
-</div>
-
```
:+1; That is a refactor

mbayopanda commented on client/src/modules/journal/journal.html:
```diff
@@ -60,13 +55,30 @@
+            class="btn btn-default" data-method="edit">
+            <i class="fa fa-pencil-square-o"></i>
+              <span ng-if="JournalCtrl.selection.selected.groups.length === 1">({{JournalCtrl.selection.selected.groups[0]}})</span>
+              <span ng-if="JournalCtrl.selection.selected.groups.length > 1">(...)</span>
```
:+1;

mbayopanda commented on client/src/modules/journal/journal.service.js:
```diff
@@ -133,5 +167,47 @@ function JournalService(Api, AppCache, Filters, Periods, Modal) {
+   * @method openCommentModal
+   * @param {object} request
+   */
+  function openCommentModal(request) {
```
Is accepted that posted transactions cannot be commented ? The main reason for the comment tool is for analysis, so i think it will be good to be able to comment also posted transactions for analysis to.

jniles replied:
Oh man, that's a bug.

mbayopanda commented on client/src/modules/journal/modals/editTransaction.modal.html:
```diff
@@ -0,0 +1,116 @@
+      </div>
+    </div>
+
+    <div class="alert alert-danger" id="validation-errored-alert" ng-if="ModalCtrl.validation.errored"  style="border-radius : 0px; padding-left : 5px; padding-right : 5px; padding-top : 10px; padding-bottom : 10px; margin-bottom : 10px;">
```
The style 
```css
border-radius : 0px;
padding-left : 5px; 
padding-right : 5px;
padding-top : 10px;
padding-bottom : 10px;
margin-bottom : 10px;
```
Can be set in the `bhima-bootstrap.less` file because it is used more than one place.

mbayopanda commented on client/src/modules/journal/modals/editTransaction.modal.js:
```diff
@@ -0,0 +1,386 @@
+        return ERROR_NEGATIVE_NUMBERS;
+      }
+
+      window.util = util;
```
What is the use of this `window.util = util` here ?

jniles replied:
Ha!  Good catch.

mbayopanda commented on client/src/modules/journal/trial-balance/errors.js:
```diff
@@ -0,0 +1,71 @@
+TrialBalanceErrorsController.$inject = ['TrialBalanceService', 'NotifyService'];
+
+/**
+ * @module TrialBalanceErrorsController
```
Is it appropriate to describe a controller as  a module ?
```js
/**
* @module TrialBalanceErrorsController
*/
```

jniles replied:
Updated to `@overview`.

mbayopanda commented on client/src/modules/journal/trial-balance/overview.js:
```diff
@@ -0,0 +1,212 @@
+        vm.loadingSubGrid = false;
+      });
+  }
+  */
```
can this commented block be removed ?

mbayopanda commented on client/src/modules/journal/trial-balance/overview.js:
```diff
@@ -0,0 +1,212 @@
+    vm.gridApi = api;
+
+    // TODO - should we really cache the trial balance transactions?
+    // api.expandable.on.rowExpandedStateChanged(null, loadSubGridRecords);
```
if the `loadSubGridRecords` function is removed, these commented lines can also be removed.

mbayopanda commented on server/controllers/finance/reports/journal/report.handlebars:
```diff
@@ -57,15 +52,15 @@
     </div>
 
     <div class="col-xs-4">
-      {{translate 'TABLE.COLUMNS.RESULT'}}: <strong>{{currency (substract (sum rows 'debit_equiv') (sum rows 'credit_equiv'))}}</strong>
```
👍

mbayopanda reviewed (approved):
This is an awesome refactoring job.

There are some non-blocking comments but this PR LGTM

jniles committed:
fix(journal): typos and commenting on posted recs

This commit fixes a few typos, consolidates a few CSS rules and removes
unused code.  It fixes a bug that disallowed posted records from being
commented, as this path was never tested before.
